File: bindings/deno/src/safe_path.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafePath - Filesystem path operations that prevent traversal attacks.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, readAndFreeString, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Safe path operations backed by formally verified Idris 2 code.
 */
export class SafePath {
  /**
   * Check if a path attempts directory traversal (e.g. "../../../etc/passwd").
   *
   * @param path - The path to check.
   * @returns Result containing a boolean or an error string.
   */
  static hasTraversal(path: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(path);
    const result = symbols.proven_path_has_traversal(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Sanitize a filename by removing dangerous characters.
   *
   * @param filename - The filename to sanitize.
   * @returns Result containing the sanitized filename or an error string.
   */
  static sanitizeFilename(filename: string): Result<string> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(filename);
    const result = symbols.proven_path_sanitize_filename(bytes, bytes.length);
    const status = result[0];
    if (status !== ProvenStatus.OK) return err(statusToError(status));
    const sanitized = readAndFreeString(result[1], Number(result[2]));
    if (sanitized === null) return err('Null string returned');
    return ok(sanitized);
  }
}

File: bindings/matlab/+proven/SafePath.m
% SPDX-License-Identifier: PMPL-1.0-or-later
% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%
% SafePath - Filesystem traversal prevention via libproven FFI.
%
% All computation is performed in the Idris 2 core via the Zig FFI bridge.
% This class is a thin wrapper; it does NOT reimplement any logic.

classdef SafePath
    % SafePath  Formally verified path safety operations.
    %
    % Methods return [] on error, or a char/logical value on success.
    %
    % Example:
    %   proven.LibProven.init();
    %   tf = proven.SafePath.hasTraversal('../../etc/passwd');
    %   % tf == true

    methods (Static)

        function result = hasTraversal(path)
            % HASTRAVERSAL  Check if path contains directory traversal.
            %
            %   tf = proven.SafePath.hasTraversal(path)
            %
            % Parameters:
            %   path - character vector.
            %
            % Returns logical true if ".." traversal detected, false if
            % safe, or [] on error.
            proven.LibProven.ensureLoaded();
            data = uint8(path);
            r = calllib(proven.LibProven.libName(), ...
                        'proven_path_has_traversal', data, numel(data));
            if r.status == 0
                result = logical(r.value);
            else
                result = [];
            end
        end

        function result = sanitizeFilename(filename)
            % SANITIZEFILENAME  Remove dangerous characters from filename.
            %
            %   safe = proven.SafePath.sanitizeFilename(filename)
            %
            % Returns sanitized char vector or [] on error.
            proven.LibProven.ensureLoaded();
            data = uint8(filename);
            r = calllib(proven.LibProven.libName(), ...
                        'proven_path_sanitize_filename', data, numel(data));
            if r.status == 0 && r.length > 0
                result = char(r.value(1:r.length));
                proven.LibProven.freeString(r.value);
            elseif r.status == 0
                result = '';
            else
                result = [];
            end
        end

    end
end
